File: scripts/migrate-blog.py
#!/usr/bin/env python3
"""
Extract blog content from downloaded reference site and create MDX files.

Architecture: Content migration script to transform HTML to structured MDX.
"""

import os
import re
import json
from pathlib import Path
from datetime import datetime

try:
    from bs4 import BeautifulSoup
except ImportError:
    print("Installing beautifulsoup4...")
    import subprocess
    subprocess.run(["pip3", "install", "beautifulsoup4"], check=True)
    from bs4 import BeautifulSoup


def slugify(text: str) -> str:
    """Convert text to URL-friendly slug."""
    text = text.lower()
    text = re.sub(r'[äÄ]', 'ae', text)
    text = re.sub(r'[öÖ]', 'oe', text)
    text = re.sub(r'[üÜ]', 'ue', text)
    text = re.sub(r'[ß]', 'ss', text)
    text = re.sub(r'[^a-z0-9]+', '-', text)
    text = text.strip('-')
    return text


def extract_post_data(html_path: Path) -> dict | None:
    """Extract blog post data from HTML file."""
    try:
        with open(html_path, 'r', encoding='utf-8') as f:
            soup = BeautifulSoup(f.read(), 'html.parser')
    except Exception as e:
        print(f"Error reading {html_path}: {e}")
        return None

    # Extract title
    title_elem = soup.select_one('h1, .blog-item-title, article h1')
    title = title_elem.get_text(strip=True) if title_elem else html_path.stem.replace('-', ' ').title()

    # Extract date
    date_elem = soup.select_one('time, .blog-meta-item--date, [datetime]')
    if date_elem:
        date_str = date_elem.get('datetime') or date_elem.get_text(strip=True)
        try:
            # Try various date formats
            for fmt in ['%Y-%m-%dT%H:%M:%S%z', '%Y-%m-%d', '%d.%m.%Y', '%d. %b %Y', '%d. %B %Y']:
                try:
                    dt = datetime.strptime(date_str.split('+')[0].split('T')[0] if 'T' in date_str else date_str, fmt.replace('T%H:%M:%S%z', ''))
                    break
                except:
                    continue
            else:
                dt = datetime.now()
        except:
            dt = datetime.now()
    else:
        dt = datetime.now()

    # Extract categories
    categories = []
    category_elems = soup.select('.blog-categories a, .blog-meta-item--categories a, [class*="category"] a')
    for cat in category_elems:
        cat_text = cat.get_text(strip=True)
        if cat_text and len(cat_text) < 50:
            categories.append(cat_text)

    # Extract excerpt/description
    meta_desc = soup.select_one('meta[name="description"]')
    excerpt = meta_desc['content'] if meta_desc and meta_desc.get('content') else ''

    if not excerpt:
        # Try first paragraph
        first_p = soup.select_one('article p, .blog-item-content p')
        if first_p:
            excerpt = first_p.get_text(strip=True)[:300]

    # Extract featured image
    og_image = soup.select_one('meta[property="og:image"]')
    featured_image = og_image['content'] if og_image and og_image.get('content') else None

    # Extract main content
    content_elem = soup.select_one('article .blog-item-content, .entry-content, article .content, main article')
    if content_elem:
        # Remove scripts, styles, nav elements
        for elem in content_elem.select('script, style, nav, header, footer, .share-buttons'):
            elem.decompose()
        content = content_elem.get_text(separator='\n\n', strip=True)
    else:
        content = ''

    # Generate slug from filename
    slug = html_path.stem

    return {
        'title': title,
        'slug': slug,
        'excerpt': excerpt[:300] if excerpt else f"Blog-Beitrag: {title}",
        'featuredImage': featured_image,
        'categories': list(set(categories))[:5],  # Dedupe and limit
        'publishedAt': dt.strftime('%Y-%m-%dT12:00:00.000Z'),
        'author': {
            'name': 'Noel Bossart',
            'bio': 'Gründer von Noevu und Experte für Webdesign und digitale Strategien.',
        },
        'content': content,
    }


def create_mdx_file(post_data: dict, output_dir: Path) -> None:
    """Create MDX file from extracted post data."""
    output_dir.mkdir(parents=True, exist_ok=True)

    filename = f"{post_data['slug']}.mdx"
    filepath = output_dir / filename

    # Create frontmatter
    frontmatter = {
        'title': post_data['title'],
        'slug': post_data['slug'],
        'excerpt': post_data['excerpt'],
        'categories': post_data['categories'],
        'author': post_data['author'],
        'publishedAt': post_data['publishedAt'],
    }

    if post_data['featuredImage']:
        frontmatter['featuredImage'] = post_data['featuredImage']

    # Build MDX content
    mdx_content = "---\n"
    for key, value in frontmatter.items():
        if isinstance(value, str):
            # Escape quotes in strings
            escaped = value.replace('"', '\\"')
            mdx_content += f'{key}: "{escaped}"\n'
        elif isinstance(value, list):
            mdx_content += f'{key}:\n'
            for item in value:
                mdx_content += f'  - "{item}"\n'
        elif isinstance(value, dict):
            mdx_content += f'{key}:\n'
            for k, v in value.items():
                if v:
                    escaped = str(v).replace('"', '\\"')
                    mdx_content += f'  {k}: "{escaped}"\n'
    mdx_content += "---\n\n"

    # Add content
    content = post_data.get('content', '')
    if content:
        # Split into paragraphs and format
        paragraphs = content.split('\n\n')
        for para in paragraphs:
            para = para.strip()
            if para:
                mdx_content += f"{para}\n\n"
    else:
        mdx_content += "Inhalt folgt...\n"

    # Write file
    with open(filepath, 'w', encoding='utf-8') as f:
        f.write(mdx_content)

    print(f"Created: {filepath.name}")


def main():
    reference_dir = Path("reference/noevu.ch/blog")
    output_dir = Path("src/content/blog")

    if not reference_dir.exists():
        print(f"Reference directory not found: {reference_dir}")
        return

    # Get all HTML files (excluding category/tag pages)
    html_files = [
        f for f in reference_dir.glob("*.html")
        if not any(x in f.name for x in ['category=', 'tag=', 'offset=', '?'])
    ]

    print(f"Found {len(html_files)} blog posts to migrate\n")

    migrated = 0
    for html_file in sorted(html_files):
        print(f"Processing: {html_file.name}")
        post_data = extract_post_data(html_file)

        if post_data and post_data.get('title'):
            create_mdx_file(post_data, output_dir)
            migrated += 1
        else:
            print(f"  Skipped: Could not extract data")

    print(f"\nMigration complete: {migrated} posts created")


if __name__ == "__main__":
    main()

File: scripts/migrate-blog.mjs
/**
 * Extract blog content from downloaded reference site and create MDX files.
 *
 * Architecture: Content migration script to transform HTML to structured MDX.
 */

import { load } from 'cheerio';
import { readFileSync, writeFileSync, readdirSync, mkdirSync, existsSync } from 'fs';
import { join, basename } from 'path';

const REFERENCE_DIR = 'reference/noevu.ch/blog';
const OUTPUT_DIR = 'src/content/blog';

function extractPostData(htmlPath) {
  let html;
  try {
    html = readFileSync(htmlPath, 'utf-8');
  } catch (error) {
    console.error(`Error reading ${htmlPath}:`, error.message);
    return null;
  }

  const $ = load(html);
  const filename = basename(htmlPath, '.html');

  // Extract title
  const titleElem = $('h1').first();
  const title = titleElem.text().trim() || filename.replace(/-/g, ' ').replace(/\b\w/g, c => c.toUpperCase());

  // Extract date from meta or schema
  let publishedAt = new Date().toISOString();
  const dateModified = $('meta[property="article:modified_time"]').attr('content');
  const datePublished = $('meta[property="article:published_time"]').attr('content');
  const timeElem = $('time').attr('datetime');

  // Helper to safely parse dates
  function tryParseDate(dateStr) {
    if (!dateStr) return null;
    try {
      const d = new Date(dateStr);
      if (isNaN(d.getTime())) return null;
      return d.toISOString();
    } catch {
      return null;
    }
  }

  publishedAt = tryParseDate(datePublished)
    || tryParseDate(dateModified)
    || tryParseDate(timeElem)
    || new Date().toISOString();

  // Extract categories
  const categories = [];
  $('a[href*="category"]').each((_, el) => {
    const cat = $(el).text().trim();
    if (cat && cat.length < 50 && !categories.includes(cat)) {
      categories.push(cat);
    }
  });

  // Extract excerpt/description
  let excerpt = $('meta[name="description"]').attr('content') || '';
  if (!excerpt) {
    const firstP = $('article p, .blog-item-content p').first().text();
    excerpt = firstP.slice(0, 300).trim();
  }

  // Extract featured image
  const featuredImage = $('meta[property="og:image"]').attr('content') || null;

  // Extract main content
  const contentElem = $('article, .blog-item-content, .entry-content').first();

  // Remove unwanted elements
  contentElem.find('script, style, nav, header, footer, .share-buttons, .blog-meta').remove();

  // Convert content to markdown-ish text
  let content = '';
  contentElem.find('p, h2, h3, h4, ul, ol, blockquote').each((_, el) => {
    const $el = $(el);
    const tag = el.tagName.toLowerCase();

    if (tag === 'h2') {
      content += `\n## ${$el.text().trim()}\n\n`;
    } else if (tag === 'h3') {
      content += `\n### ${$el.text().trim()}\n\n`;
    } else if (tag === 'h4') {
      content += `\n#### ${$el.text().trim()}\n\n`;
    } else if (tag === 'ul' || tag === 'ol') {
      $el.find('li').each((_, li) => {
        content += `- ${$(li).text().trim()}\n`;
      });
      content += '\n';
    } else if (tag === 'blockquote') {
      content += `> ${$el.text().trim()}\n\n`;
    } else {
      const text = $el.text().trim();
      if (text && text.length > 10) {
        content += `${text}\n\n`;
      }
    }
  });

  return {
    title,
    slug: filename,
    excerpt: excerpt || `Blog-Beitrag: ${title}`,
    featuredImage,
    categories: categories.slice(0, 5),
    publishedAt,
    author: {
      name: 'Noel Bossart',
      bio: 'Gründer von Noevu und Experte für Webdesign und digitale Strategien.',
    },
    content: content.trim() || 'Inhalt folgt...',
  };
}

function escapeYaml(str) {
  if (!str) return '';
  // Escape special characters for YAML strings
  return str.replace(/"/g, '\\"').replace(/\n/g, ' ');
}

function createMdxFile(postData, outputDir) {
  if (!existsSync(outputDir)) {
    mkdirSync(outputDir, { recursive: true });
  }

  const filename = `${postData.slug}.mdx`;
  const filepath = join(outputDir, filename);

  // Build frontmatter
  let mdx = '---\n';
  mdx += `title: "${escapeYaml(postData.title)}"\n`;
  mdx += `slug: "${postData.slug}"\n`;
  mdx += `excerpt: "${escapeYaml(postData.excerpt)}"\n`;

  if (postData.featuredImage) {
    mdx += `featuredImage: "${postData.featuredImage}"\n`;
  }

  if (postData.categories.length > 0) {
    mdx += 'categories:\n';
    postData.categories.forEach(cat => {
      mdx += `  - "${escapeYaml(cat)}"\n`;
    });
  }

  mdx += 'author:\n';
  mdx += `  name: "${postData.author.name}"\n`;
  if (postData.author.bio) {
    mdx += `  bio: "${escapeYaml(postData.author.bio)}"\n`;
  }

  mdx += `publishedAt: "${postData.publishedAt}"\n`;
  mdx += '---\n\n';

  // Add content
  mdx += postData.content;

  writeFileSync(filepath, mdx, 'utf-8');
  console.log(`Created: ${filename}`);
}

function main() {
  if (!existsSync(REFERENCE_DIR)) {
    console.error(`Reference directory not found: ${REFERENCE_DIR}`);
    process.exit(1);
  }

  // Get all HTML files
  const htmlFiles = readdirSync(REFERENCE_DIR)
    .filter(f => f.endsWith('.html'))
    .filter(f => !f.includes('category=') && !f.includes('tag=') && !f.includes('offset=') && !f.includes('?'));

  console.log(`Found ${htmlFiles.length} blog posts to migrate\n`);

  let migrated = 0;
  for (const file of htmlFiles.sort()) {
    const htmlPath = join(REFERENCE_DIR, file);
    console.log(`Processing: ${file}`);

    const postData = extractPostData(htmlPath);
    if (postData && postData.title) {
      createMdxFile(postData, OUTPUT_DIR);
      migrated++;
    } else {
      console.log('  Skipped: Could not extract data');
    }
  }

  console.log(`\nMigration complete: ${migrated} posts created`);
}

main();
